#!/bin/bash

################################################################################################ 
	#################### Torque preparation: PBS commands ###################
################################################################################################

### Use the bourne shell
#PBS -S /bin/bash 			

### Run in the queue named default
#PBS -q default				

#### #PBS -d /home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/src-azza

### Specify the number of cpus for your job (nodes and processors)
#PBS -l nodes=1:ppn=23			

### the destination for reporting (defaults to script's directory): Seperate these to get a better sense!
#PBS -o localhost:$HOME/outputs.log
#PBS -e localhost:$HOME/errors.log		

### email me when my job aborts or ends
#PBS -m abe
#PBS -M azzaea@gamil.com



# Calculate the number of processors allocated to this run.
NPROCS=`wc -l < $PBS_NODEFILE`

# Calculate the number of nodes allocated.
NNODES=`uniq $PBS_NODEFILE | wc -l`

### Display the job context
echo
echo "Running the job named ($PBS_JOBNAME) identified by ($PBS_JOBID) requested by ($PBS_O_LOGNAME)" >> jobinfo.txt
echo "Job is run on host (`hostname`) from the host cluster ($PBS_O_HOST)" >> jobinfo.txt
echo Using ${NPROCS} processors across ${NNODES} nodes >> jobinfo.txt
echo -e  "\n Time is `date`" >> jobinfo.txt

### By default TORQUE launches processes from your home directory, you may need to switch to the working directory; 
printf "\n Note: This job is being executed on the directory `pwd`, even though it was submitted from the directory $PBS_O_WORKDIR "


echo -e "\n\n########################################################################################"
echo -e "#############                Pipeline starts here!              ###############"
echo -e "########################################################################################\n\n"

######### Paths defintions:
reference="/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/data/genome"
reads="/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/data/reads"
results="/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/results"
	# an alternative to having to copy the reference genome (from the gatk bundle in the mirror folder), would have been creating a soft link to it:
	# ln -s /home/mirrors/gatk_bundle/2.8/hg19/ucsc.hg19.fasta reference
	# and now from this point onward, reference is accepted instead of the long path to ucsc.hg19.fasta

######### Some parts were already done, so it would be a good idea to not rerun them!
if [ $done_parts ]; then

######### Reference preparation:
gunzip $reference/ucsc*

# The used reference is from the gatk bundle, so no need to do the steps below:
	#module load samtools/1.2
	#samtools faidx $reads/hs_ref_GRCh38.p2_chr1.fa
# Actually you need to index. The indexed file from gatk was generated by fadix, not by bwa's index!
module load bwa/0.7.10
cd $reference 
# bwa index is incapable of moving its output around. Do it manually!Generally speaking(moving the files from indexing can be tricky with some tools. Better to always go the directory)
bwa index -a bwtsw -p human $reference/ucsc.hg19.fasta
	# the -p parameter is just to name the resulting files human.xx where xx is the various extensions!

# Also note you are using the complete genome even though you are only interested in certain parts
# The good news is that this is a one time thing! Your reference can now deal with whatever it is that is human!

######### Quality checking: 
module load fastqc/0.11.4
fastqc $reads/H3A_NextGen_assessment.Chr1_50X.set3_read1.fq --outdir=$results
fastqc $reads/H3A_NextGen_assessment.Chr1_50X.set3_read2.fq --outdir=$results
# In reality, to see this result, just type: firefox required.read_fastqc.html
# Reports show data of good quality, except for some kmer content in read1 (Gloria: its near the end, so should be ok). Also, the encoding is Sanger, so the reads are neat..
# A pertinent question is about the read group, Do I accept the info presented on the chromosome here?


######### for the sake of sanity, I created 2 couple read files:
cp H3A_NextGen_assessment.Chr1_50X.set3_read1.fq read1.fq
cp H3A_NextGen_assessment.Chr1_50X.set3_read2.fq read2.fq


######### Alignment
module load bwa/0.7.10
bwa mem -M -R  '@RG\tID:foo\tSM:bar\tLB:library1' $reference/human $reads/read1.fq $reads/read2.fq > $results/p2_alignment/a1.default.sam
# note that it is the indexed file(s) that was needed in this stage. It has a different name than the reference (remember the -p argument), so I need to use its name (human)


######### Making sense of the Alignment
module load samtools/1.2

samtools flagstat $results/p2_alignment/a1.default.sam > $results/p2_alignment/a1.default_summary.txt

samtools view -bS $results/p2_alignment/a1.default.sam > $results/p2_alignment/a1.default.bam
fi

module load bwa/0.7.10
module load samtools/1.2
cd $results/p2_alignment/
#for (i=0; i<=30; i=i+3); do
for i in $(seq 10 2 30); do
	START=$(date +%s)	
	bwa mem -k "$i" -M -R  '@RG\tID:foo\tSM:bar\tLB:library1'  $reference/human $reads/read1.fq $reads/read2.fq > "a$i.sam"
	samtools view -bS a$i.sam > "a$i.bam"
	END=$(date +%s)
	DIFF=$(( $END - $START ))
	echo 
	echo "BWA Mem using parameter -k = $i" > "a$i.summary.txt"
	echo 
	echo "Execution time is $DIFF seconds" >> "a$i.summary.txt"
	echo 
	samtools flagstat a$i.sam >> "a$i.summary.txt"
done


#igv mAYBE





#module load picard-tools/1.141

# I actually ran this command on my node itself!!!
#java -jar picard.jar FastqToSam F1=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/data/reads/H3A_NextGen_assessment.Chr1_50X.set3_read1.fq F2=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/data/reads/H3A_NextGen_assessment.Chr1_50X.set3_read2.fq   O=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/results/fastq_to_bam.bam     SM=for_tool_testing 

#java -jar picard.jar MergeBamAlignment ALIGNED=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/results/aligned.bam UNMAPPED=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/results/fastq_to_bam.bam    O=merge_alignments.bam      R=/home/groups/hpcbio_shared/azza/H3A_NextGen_assessment_set3/data/reads/hs_ref_GRCh38.p2_chr1.fa

#################################################################################################### 
################################ What are the group read info??? ################################
####################################################################################################


echo -e "\n\n########################################################################################"
echo -e "##############                 EXITING NOW                            ##################"	
echo -e "########################################################################################\n\n"
